File: sdks/python/src/untrace/tracer.py
"""Tracer implementation for the Untrace SDK."""

from typing import Any, Callable, Dict, Optional, Union
from opentelemetry import context, trace
from opentelemetry.trace import Span, SpanKind, StatusCode, Tracer
from opentelemetry.context import Context

from .types import UntraceSpanOptions, LLMSpanAttributes


class UntraceTracer:
    """Wrapper around OpenTelemetry tracer with LLM-specific helpers."""

    def __init__(self, tracer: Tracer):
        """Initialize the Untrace tracer.

        Args:
            tracer: OpenTelemetry tracer instance
        """
        self._tracer = tracer

    def get_tracer(self) -> Tracer:
        """Get the underlying OpenTelemetry tracer.

        Returns:
            OpenTelemetry tracer instance
        """
        return self._tracer

    def start_span(self, options: UntraceSpanOptions) -> Span:
        """Start a new span.

        Args:
            options: Span options

        Returns:
            Started span
        """
        span_options = {
            "attributes": options.attributes,
            "kind": options.kind,
        }

        parent_context = options.parent
        if parent_context is None:
            parent_context = context.get_current()

        return self._tracer.start_span(
            options.name,
            context=parent_context,
            **span_options
        )

    def start_llm_span(
        self,
        name: str,
        attributes: Union[LLMSpanAttributes, Dict[str, Any]],
        parent: Optional[Span] = None,
    ) -> Span:
        """Start an LLM span.

        Args:
            name: Span name
            attributes: LLM attributes
            parent: Parent span

        Returns:
            Started LLM span
        """
        if isinstance(attributes, LLMSpanAttributes):
            attrs = attributes.to_dict()
        else:
            attrs = attributes

        return self.start_span(
            UntraceSpanOptions(
                name=name,
                attributes=attrs,
                kind=SpanKind.CLIENT,
                parent=parent,
            )
        )

    def start_workflow_span(
        self,
        name: str,
        attributes: Optional[Dict[str, Any]] = None,
        parent: Optional[Span] = None,
    ) -> Span:
        """Start a workflow span.

        Args:
            name: Span name
            attributes: Span attributes
            parent: Parent span

        Returns:
            Started workflow span
        """
        attrs = {"workflow.name": name}
        if attributes:
            attrs.update(attributes)

        return self.start_span(
            UntraceSpanOptions(
                name=name,
                attributes=attrs,
                kind=SpanKind.INTERNAL,
                parent=parent,
            )
        )

    def start_tool_span(
        self,
        name: str,
        attributes: Optional[Dict[str, Any]] = None,
        parent: Optional[Span] = None,
    ) -> Span:
        """Start a tool span.

        Args:
            name: Span name
            attributes: Span attributes
            parent: Parent span

        Returns:
            Started tool span
        """
        attrs = {"tool.name": name}
        if attributes:
            attrs.update(attributes)

        return self.start_span(
            UntraceSpanOptions(
                name=name,
                attributes=attrs,
                kind=SpanKind.INTERNAL,
                parent=parent,
            )
        )

    async def with_span(
        self,
        options: UntraceSpanOptions,
        fn: Callable[[Span], Any],
    ) -> Any:
        """Wrap a function with a span.

        Args:
            options: Span options
            fn: Function to wrap

        Returns:
            Function result
        """
        span = self.start_span(options)

        try:
            result = await fn(span)
            span.set_status(StatusCode.OK)
            return result
        except Exception as error:
            span.record_exception(error)
            span.set_status(
                StatusCode.ERROR,
                description=str(error) if error else "Unknown error"
            )
            raise
        finally:
            span.end()

    async def with_llm_span(
        self,
        name: str,
        attributes: Union[LLMSpanAttributes, Dict[str, Any]],
        fn: Callable[[Span], Any],
    ) -> Any:
        """Wrap an LLM call with a span.

        Args:
            name: Span name
            attributes: LLM attributes
            fn: Function to wrap

        Returns:
            Function result
        """
        if isinstance(attributes, LLMSpanAttributes):
            attrs = attributes.to_dict()
        else:
            attrs = attributes

        return await self.with_span(
            UntraceSpanOptions(
                name=name,
                attributes=attrs,
                kind=SpanKind.CLIENT,
            ),
            fn,
        )

    def get_active_span(self) -> Optional[Span]:
        """Get the active span.

        Returns:
            Active span or None
        """
        return trace.get_current_span()

    def add_event(self, name: str, attributes: Optional[Dict[str, Any]] = None) -> None:
        """Add event to the active span.

        Args:
            name: Event name
            attributes: Event attributes
        """
        span = self.get_active_span()
        if span:
            span.add_event(name, attributes or {})

    def set_attributes(self, attributes: Dict[str, Any]) -> None:
        """Set attributes on the active span.

        Args:
            attributes: Attributes to set
        """
        span = self.get_active_span()
        if span:
            span.set_attributes(attributes)

    def set_status(self, code: StatusCode, message: Optional[str] = None) -> None:
        """Set status on the active span.

        Args:
            code: Status code
            message: Status message
        """
        span = self.get_active_span()
        if span:
            span.set_status(code, description=message)

    def record_exception(
        self,
        error: Exception,
        attributes: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Record an exception on the active span.

        Args:
            error: Exception to record
            attributes: Additional attributes
        """
        span = self.get_active_span()
        if span:
            if attributes:
                span.set_attributes(attributes)
            span.record_exception(error)

File: sdks/go/untrace/tracer.go
package untrace

import (
	"context"
	"fmt"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"
)

// untraceTracer implements the Tracer interface
type untraceTracer struct {
	tracer trace.Tracer
}

// NewTracer creates a new Untrace tracer
func NewTracer(tracer trace.Tracer) Tracer {
	return &untraceTracer{
		tracer: tracer,
	}
}

// StartLLMSpan starts a new LLM span with appropriate attributes
func (t *untraceTracer) StartLLMSpan(ctx context.Context, name string, opts LLMSpanOptions) (context.Context, trace.Span) {
	attrs := t.buildLLMAttributes(opts)

	spanCtx, span := t.tracer.Start(ctx, name, trace.WithAttributes(attrs...))

	return spanCtx, span
}

// StartSpan starts a new span with the given options
func (t *untraceTracer) StartSpan(ctx context.Context, name string, opts SpanOptions) (context.Context, trace.Span) {
	var spanOpts []trace.SpanStartOption

	if opts.Kind != trace.SpanKindInternal {
		spanOpts = append(spanOpts, trace.WithSpanKind(opts.Kind))
	}

	if opts.Parent.IsValid() {
		spanCtx := trace.ContextWithSpanContext(ctx, opts.Parent)
		ctx = spanCtx
	}

	attrs := t.buildAttributes(opts.Attributes)
	if len(attrs) > 0 {
		spanOpts = append(spanOpts, trace.WithAttributes(attrs...))
	}

	spanCtx, span := t.tracer.Start(ctx, name, spanOpts...)
	return spanCtx, span
}

// GetTracer returns the underlying OpenTelemetry tracer
func (t *untraceTracer) GetTracer() trace.Tracer {
	return t.tracer
}

// buildLLMAttributes builds attributes for LLM spans
func (t *untraceTracer) buildLLMAttributes(opts LLMSpanOptions) []attribute.KeyValue {
	attrs := []attribute.KeyValue{
		attribute.String("llm.provider", opts.Provider),
		attribute.String("llm.model", opts.Model),
		attribute.String("llm.operation.type", string(opts.Operation)),
	}

	if opts.PromptTokens != nil {
		attrs = append(attrs, attribute.Int("llm.prompt.tokens", *opts.PromptTokens))
	}
	if opts.CompletionTokens != nil {
		attrs = append(attrs, attribute.Int("llm.completion.tokens", *opts.CompletionTokens))
	}
	if opts.TotalTokens != nil {
		attrs = append(attrs, attribute.Int("llm.total.tokens", *opts.TotalTokens))
	}
	if opts.Temperature != nil {
		attrs = append(attrs, attribute.Float64("llm.temperature", *opts.Temperature))
	}
	if opts.TopP != nil {
		attrs = append(attrs, attribute.Float64("llm.top_p", *opts.TopP))
	}
	if opts.MaxTokens != nil {
		attrs = append(attrs, attribute.Int("llm.max_tokens", *opts.MaxTokens))
	}
	if opts.Stream != nil {
		attrs = append(attrs, attribute.Bool("llm.stream", *opts.Stream))
	}
	if opts.Tools != nil {
		attrs = append(attrs, attribute.String("llm.tools", *opts.Tools))
	}
	if opts.ToolCalls != nil {
		attrs = append(attrs, attribute.String("llm.tool_calls", *opts.ToolCalls))
	}
	if opts.DurationMs != nil {
		attrs = append(attrs, attribute.Int("llm.duration_ms", *opts.DurationMs))
	}
	if opts.CostPrompt != nil {
		attrs = append(attrs, attribute.Float64("llm.cost.prompt", *opts.CostPrompt))
	}
	if opts.CostCompletion != nil {
		attrs = append(attrs, attribute.Float64("llm.cost.completion", *opts.CostCompletion))
	}
	if opts.CostTotal != nil {
		attrs = append(attrs, attribute.Float64("llm.cost.total", *opts.CostTotal))
	}
	if opts.Error != nil {
		attrs = append(attrs, attribute.String("llm.error", *opts.Error))
	}
	if opts.ErrorType != nil {
		attrs = append(attrs, attribute.String("llm.error.type", *opts.ErrorType))
	}
	if opts.RequestID != nil {
		attrs = append(attrs, attribute.String("llm.request.id", *opts.RequestID))
	}
	if opts.UsageReason != nil {
		attrs = append(attrs, attribute.String("llm.usage.reason", *opts.UsageReason))
	}

	// Add custom attributes
	customAttrs := t.buildAttributes(opts.Attributes)
	attrs = append(attrs, customAttrs...)

	return attrs
}

// buildAttributes converts a map of attributes to OpenTelemetry attributes
func (t *untraceTracer) buildAttributes(attrs map[string]interface{}) []attribute.KeyValue {
	var result []attribute.KeyValue

	for key, value := range attrs {
		switch v := value.(type) {
		case string:
			result = append(result, attribute.String(key, v))
		case int:
			result = append(result, attribute.Int(key, v))
		case int64:
			result = append(result, attribute.Int64(key, v))
		case float64:
			result = append(result, attribute.Float64(key, v))
		case bool:
			result = append(result, attribute.Bool(key, v))
		case []string:
			result = append(result, attribute.StringSlice(key, v))
		case []int:
			result = append(result, attribute.IntSlice(key, v))
		case []float64:
			result = append(result, attribute.Float64Slice(key, v))
		case time.Time:
			result = append(result, attribute.String(key, v.Format(time.RFC3339)))
		default:
			// Convert to string as fallback
			result = append(result, attribute.String(key, fmt.Sprintf("%v", v)))
		}
	}

	return result
}
